File: src/CircleAI.Telephony/LocalDevTunnel.cs
// LocalDevTunnel.cs
//
// (3.3.0) Local-dev tunnel resolver. A voice loop needs an
// internet-reachable webhook URL even when running locally. This
// abstraction lets dev configurations route through Cloudflare
// Tunnel, ngrok, or a manually-pinned static URL — same interface,
// different backing.

using System;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Telephony;

/// <summary>(3.3.0) Resolves a public, internet-reachable URL that maps to a local port.</summary>
public interface ILocalDevTunnel
{
    /// <summary>Identifier — "cloudflare", "ngrok", "static", "null".</summary>
    string ProviderId { get; }

    /// <summary>Whether this resolver is configured/available.</summary>
    bool IsAvailable { get; }

    /// <summary>Resolve the public URL forwarding to <paramref name="localPort"/>.</summary>
    ValueTask<Uri> GetPublicUrlAsync(int localPort, CancellationToken ct = default);
}

/// <summary>(3.3.0) DI-default that throws — host wires a real tunnel.</summary>
public sealed class NullLocalDevTunnel : ILocalDevTunnel
{
    public static readonly NullLocalDevTunnel Instance = new();
    public string ProviderId   => "null";
    public bool   IsAvailable  => false;
    public ValueTask<Uri> GetPublicUrlAsync(int localPort, CancellationToken ct = default)
        => throw new InvalidOperationException(
            "No local-dev tunnel is configured. Register a CloudflareTunnel / NgrokTunnel / StaticTunnel.");
}

/// <summary>(3.3.0) Static-URL tunnel — caller supplies the public URL up front (best for CI).</summary>
public sealed class StaticLocalDevTunnel : ILocalDevTunnel
{
    private readonly Uri _publicUrl;

    public StaticLocalDevTunnel(Uri publicUrl)
    {
        _publicUrl = publicUrl ?? throw new ArgumentNullException(nameof(publicUrl));
        if (!publicUrl.IsAbsoluteUri) throw new ArgumentException("publicUrl must be absolute.", nameof(publicUrl));
    }

    public string ProviderId   => "static";
    public bool   IsAvailable  => true;
    public ValueTask<Uri> GetPublicUrlAsync(int localPort, CancellationToken ct = default)
        => ValueTask.FromResult(_publicUrl);
}

/// <summary>(3.3.0) Cloudflare Tunnel resolver. Host must point at the cloudflared output URL.</summary>
public sealed class CloudflareTunnel : ILocalDevTunnel
{
    private readonly Func<int, CancellationToken, ValueTask<Uri>> _resolver;

    public CloudflareTunnel(Func<int, CancellationToken, ValueTask<Uri>> resolver)
    {
        _resolver = resolver ?? throw new ArgumentNullException(nameof(resolver));
    }

    public string ProviderId   => "cloudflare";
    public bool   IsAvailable  => true;
    public ValueTask<Uri> GetPublicUrlAsync(int localPort, CancellationToken ct = default)
        => _resolver(localPort, ct);
}

/// <summary>(3.3.0) ngrok tunnel resolver.</summary>
public sealed class NgrokTunnel : ILocalDevTunnel
{
    private readonly Func<int, CancellationToken, ValueTask<Uri>> _resolver;

    public NgrokTunnel(Func<int, CancellationToken, ValueTask<Uri>> resolver)
    {
        _resolver = resolver ?? throw new ArgumentNullException(nameof(resolver));
    }

    public string ProviderId   => "ngrok";
    public bool   IsAvailable  => true;
    public ValueTask<Uri> GetPublicUrlAsync(int localPort, CancellationToken ct = default)
        => _resolver(localPort, ct);
}

File: python/src/circle_ai/telephony/local_dev_tunnel.py
# local_dev_tunnel.py
#
# Port of CircleAI.Telephony LocalDevTunnel.cs (C# — the EXACT spec).
#
# (3.3.0) Local-dev tunnel resolver. A voice loop needs an internet-reachable
# webhook URL even when running locally. This abstraction lets dev
# configurations route through Cloudflare Tunnel, ngrok, or a manually-pinned
# static URL — same interface, different backing.
#
# C# Uri -> str (the IsAbsoluteUri guard maps to a scheme+netloc check). C#
# Func<int, CancellationToken, ValueTask<Uri>> resolver -> an async Callable. The
# static ``Instance`` singleton on NullLocalDevTunnel maps to a module-level
# singleton. C# InvalidOperationException -> RuntimeError.

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Awaitable, Callable, Optional
from urllib.parse import urlsplit

# C# Func<int, CancellationToken, ValueTask<Uri>>.
TunnelResolver = Callable[[int, Optional[object]], Awaitable[str]]


def _is_absolute_uri(uri: str) -> bool:
    parts = urlsplit(uri)
    return bool(parts.scheme) and bool(parts.netloc)


class ILocalDevTunnel(ABC):
    """(3.3.0) Resolves a public, internet-reachable URL that maps to a local port."""

    @property
    @abstractmethod
    def provider_id(self) -> str:
        """Identifier — "cloudflare", "ngrok", "static", "null"."""

    @property
    @abstractmethod
    def is_available(self) -> bool:
        """Whether this resolver is configured/available."""

    @abstractmethod
    async def get_public_url_async(self, local_port: int, *, ct: Optional[object] = None) -> str:
        """Resolve the public URL forwarding to ``local_port``."""


class NullLocalDevTunnel(ILocalDevTunnel):
    """(3.3.0) DI-default that throws — host wires a real tunnel."""

    Instance: "NullLocalDevTunnel"

    @property
    def provider_id(self) -> str:
        return "null"

    @property
    def is_available(self) -> bool:
        return False

    async def get_public_url_async(self, local_port: int, *, ct: Optional[object] = None) -> str:
        raise RuntimeError(
            "No local-dev tunnel is configured. Register a CloudflareTunnel / NgrokTunnel / StaticTunnel."
        )


class StaticLocalDevTunnel(ILocalDevTunnel):
    """(3.3.0) Static-URL tunnel — caller supplies the public URL up front (best for CI)."""

    def __init__(self, public_url: str) -> None:
        if public_url is None:
            raise ValueError("public_url must not be None")
        if not _is_absolute_uri(public_url):
            raise ValueError("publicUrl must be absolute.")
        self._public_url = public_url

    @property
    def provider_id(self) -> str:
        return "static"

    @property
    def is_available(self) -> bool:
        return True

    async def get_public_url_async(self, local_port: int, *, ct: Optional[object] = None) -> str:
        return self._public_url


class CloudflareTunnel(ILocalDevTunnel):
    """(3.3.0) Cloudflare Tunnel resolver. Host must point at the cloudflared output URL."""

    def __init__(self, resolver: TunnelResolver) -> None:
        if resolver is None:
            raise ValueError("resolver must not be None")
        self._resolver = resolver

    @property
    def provider_id(self) -> str:
        return "cloudflare"

    @property
    def is_available(self) -> bool:
        return True

    async def get_public_url_async(self, local_port: int, *, ct: Optional[object] = None) -> str:
        return await self._resolver(local_port, ct)


class NgrokTunnel(ILocalDevTunnel):
    """(3.3.0) ngrok tunnel resolver."""

    def __init__(self, resolver: TunnelResolver) -> None:
        if resolver is None:
            raise ValueError("resolver must not be None")
        self._resolver = resolver

    @property
    def provider_id(self) -> str:
        return "ngrok"

    @property
    def is_available(self) -> bool:
        return True

    async def get_public_url_async(self, local_port: int, *, ct: Optional[object] = None) -> str:
        return await self._resolver(local_port, ct)


NullLocalDevTunnel.Instance = NullLocalDevTunnel()
